### DISCUSSION TITLE
How to assign the heterogeneous permeability coefficient  K in Nested mesh problem (DISV)?

### DISCUSSION THREAD

**Discussion by tju999** [modflowpy/flopy#2039] [Q&A] [answered]
Hi,
I'm sorry to call you, but I have some questions about DISV.I didn't find a suitable documentation case.

I follow this case '5. Nested Grid Problem':
https://modflow6-examples.readthedocs.io/en/master/_examples/ex-gwf-u1disv.html

The code is as follows:
https://modflow6-examples.readthedocs.io/en/master/_notebooks/ex-gwf-u1disv.html

My question is how can I generate a model that assigns a different permeability coefficient K,

For example, in the standard grid DIS, assuming that my study area is 64*64 layer, i.e. k11(1,64,64), I can operate with K=k11 in the permeability assignment, in which case k11 can be heterogeneous.

However, in DISV, because the grid contains nested grids, I can't directly assign values with a matrix, and of course I don't want my permeability coefficient to be just a constant C, so what should I do? I failed after some attempts and did not see the proper documentation (different assignment of permeability coefficient in DISV).

Hope to get your help, or documentation.

Looking forward to receiving your reply.

**Comment by dbrakenhoff** [comment]
Hi @tju999,

Perhaps take a look at some of the utilities for intersecting your grid with shapes or rasters. The former will allow you to identify which cells lie within a particular zone and specify different k-values for different zones. With rasters you can resample your raster containing k-values to your particular grid. 

* [Notebook on intersecting grids and shapes](https://flopy.readthedocs.io/en/latest/Notebooks/grid_intersection_example.html)
* [Notebook on grids and rasters](https://flopy.readthedocs.io/en/latest/Notebooks/raster_intersection_example.html)

Hope that helps!

**Reply by tju999** [reply]
Thank you very much for your reply,

I read the document you recommended, it is really helpful, but I still have some questions to ask.

Follow the method you recommend for your document: Select a specific grid first, then assign values using the raster data.

I'm sorry that I didn't explain it clearly, but my main research direction is groundwater simulation based on deep learning methods. At present, the permeability coefficient data I used is still randomly generated following a spatial field distribution, and I hope that each grid has its unique permeability coefficient attribute value, rather than through similar grid interpolation/assignment. Moreover, the format of the permeability coefficient field I generated is ndarray. If  follow the documented method above, it is equivalent to ndarray-raster-ndarray and this loop increases the time. Is there a  way to assign the permeability coefficient k directly to the unstructured refined mesh DISV? Because it is convenient to assign the permeability factor k directly to the matrix in the annotated structured grid DIS method, do you know of a documented method for DISV in this regard?

Looking forward to receiving your reply.

**Reply by dbrakenhoff** [reply]
The ndarray structure for DISV is (no. of layers, no. of cells per layer), so you can just pass in a random array with that shape.

**Reply by tju999** [reply]
Thank you very much for your reply. As you said, I tried to assign a value to k in disv(no. of layers, no. of cells per layer), but it failed.

Take "5. Nested Grid Problem" for example:

In the original code the assignment is passed：`k11 = 1.0     # Horizontal hydraulic conductivity ($m/d$)`.At this point everything is normal and can run

But when I used the simplest example, I also set the permeability coefficient to 1, but entered it in each grid format, but there was a problem with the code:
```
k11= []
k11+= [[0,i,1.0] for i in range(0,121)]
```
Error prompt:
```
MFDataException: An error occurred in data element "k" model "ex-gwf-u1disv" package "npf". The error occurred while setting data in the "__init__" method.
Additional Information:
(1) Unable to set data "k" layer 0.  Data is not in a valid format.
(2) Error occurred while adding dataset "k" to block "griddata"
```
My question is does the permeability coefficient not conform to the setting of this format (layer, node, value)?But both CHD and WEL packages can succeed because of this setting in disv(layer, node, value).

Can you point out what went wrong with my permeability assignment? Of course, the assignment of all 1 here is just to simplify the problem, and  assign different values in the future.

Looking forward to receiving your reply.

**Reply by dbrakenhoff** [reply, marked as answer]
The example problem has a single layer as far as I can tell, so your k-list or array should have the following shape:
```python
k = np.ones((1, 121))
or 
k = [[<121 k-values ...>]]
```
You're trying to add a list of 3 k11-values for each cell in the layer, which is why you're getting an error.

**Reply by tju999** [reply]
It works well!

Thank you very much!

I love flopy!!!
